src/pdm/cli/commands/list.py

```diff
@@ -1,37 +1,428 @@
+                for dist in working_set.values()
+            ),
+            key=lambda x: x.lower(),
+        )
+        project.core.ui.echo("\n".join(requirements))
+
+    def handle_graph(
+        self,
+        dep_graph: DirectedGraph[Package | None],
+        project: Project,
+        options: argparse.Namespace,
+    ) -> None:
+        if options.csv:
+            raise PdmUsageError("--csv cannot be used with --graph")
+        if options.markdown:
+            raise PdmUsageError("--markdown cannot be used with --graph")
+        if options.sort:
+            raise PdmUsageError("--sort cannot be used with --graph")
+
+        show_dependency_graph(
+            project, dep_graph, reverse=options.reverse, json=options.json
+        )
+
+    def handle_list(
+        self,
+        packages: Mapping[str, im.Distribution],
+        name_to_groups: Mapping[str, set[str]],
+        project: Project,
+        options: argparse.Namespace,
+    ) -> None:
+        if options.reverse:
+            raise PdmUsageError("--reverse cannot be used without --graph")
+
+        # Check the fields are specified OK.
+        fields = parse_comma_separated_string(
+            options.fields, asterisk_values=Listable.KEYS
         )
+        if not all(field in Listable.KEYS for field in fields):
+            raise PdmUsageError(
+                f"--fields must specify one or more of: {Listable.KEYS}"
+            )
+
+        # Wrap each distribution with a Listable (and a groups pairing)
+        # to make it easier to filter on later.
+        def _group_of(name: str) -> set[str]:
+            return name_to_groups.get(name, {SUBDEP_GROUP_LABEL})
+
+        records = [Listable(d, _group_of(k)) for k, d in packages.items()]
+        ui = project.core.ui
+
+        # Order based on a field key.
+        if options.sort:
+            keys = parse_comma_separated_string(options.sort)
+            if not all(key in Listable.KEYS for key in keys):
+                raise PdmUsageError(
+                    f"--sort key must be one of: {','.join(Listable.KEYS)}"
+                )
+            records.sort(key=lambda d: tuple(d[key] for key in keys))
+
+        # Write CSV
+        if options.csv:
+            buffer = io.StringIO()
+            writer = csv.DictWriter(buffer, fieldnames=fields)
+            writer.writeheader()
+            for row in records:
+                writer.writerow(row.json(fields))
+            ui.echo(buffer.getvalue(), highlight=True, end="")
+
+        # Write JSON
+        elif options.json:
+            json_row = [row.json(fields) for row in records]
+            ui.echo(json.dumps(json_row, indent=4), highlight=True)
+
+        # Write Markdown
+        elif options.markdown:
+            body = [f"# {project.name} licenses"]
+            body.extend(row.markdown(fields) for row in records)
+            ui.echo("\n".join(body), highlight=True)
```

Comment:
A few licences had unusual characters that resulted in unicode errors when printed to a non-tty, for instance the `Arpeggio` package. https://github.com/textX/Arpeggio/blob/master/LICENSE 

What will `ui.echo` do with the character?  As it was reproducing a license document, I felt it was best to add a warning to the document that said that some characters were changed. 

Ref: https://rich.readthedocs.io/en/stable/console.html#printing

Reply:
It seems both `ui.echo` and `print` work on my machine, maybe the terminal encoding is fine for me.

Reply:
Try piping the output to a file to trigger the error (powershell, windows terminal):

**pdm list --fields name,groups,licenses --markdown -v > test.md**

Here is the stack trace:
```
PS C:\Users\John\Desktop\pdm> pdm run pdm list --fields name,groups,licenses --markdown -v > test.md
Traceback (most recent call last):
  File "C:\Users\John\AppData\Local\Programs\Python\Python39\lib\runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Users\John\AppData\Local\Programs\Python\Python39\lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "C:\Users\John\Desktop\pdm\__pypackages__\3.9\Scripts\pdm.EXE\__main__.py", line 7, in <module>
  File "C:\Users\John\Desktop\pdm\src\pdm\core.py", line 258, in main
    return Core().main(args)
  File "C:\Users\John\Desktop\pdm\src\pdm\core.py", line 190, in main
    raise cast(Exception, err).with_traceback(traceback)
  File "C:\Users\John\Desktop\pdm\src\pdm\core.py", line 185, in main
    f(project, options)
  File "C:\Users\John\Desktop\pdm\src\pdm\cli\commands\list.py", line 187, in handle
    self.handle_list(packages, name_to_groups, project, options)
  File "C:\Users\John\Desktop\pdm\src\pdm\cli\commands\list.py", line 291, in handle_list
    ui.echo("\n".join(body), highlight=True)
  File "C:\Users\John\Desktop\pdm\src\pdm\termui.py", line 173, in echo
    console.print(message, **kwargs)
  File "C:\Users\John\Desktop\pdm\__pypackages__\3.9\lib\rich\console.py", line 1694, in print
    self._buffer.extend(new_segments)
  File "C:\Users\John\Desktop\pdm\__pypackages__\3.9\lib\rich\console.py", line 848, in __exit__
    self._exit_buffer()
  File "C:\Users\John\Desktop\pdm\__pypackages__\3.9\lib\rich\console.py", line 806, in _exit_buffer
    self._check_buffer()
  File "C:\Users\John\Desktop\pdm\__pypackages__\3.9\lib\rich\console.py", line 1999, in _check_buffer
    legacy_windows_render(
  File "C:\Users\John\Desktop\pdm\__pypackages__\3.9\lib\rich\_windows_renderer.py", line 19, in legacy_windows_render
    term.write_text(text)
  File "C:\Users\John\Desktop\pdm\__pypackages__\3.9\lib\rich\_win32_console.py", line 403, in write_text
    self.write(text)
  File "C:\Users\John\AppData\Local\Programs\Python\Python39\lib\encodings\cp1252.py", line 19, in encode
    return codecs.charmap_encode(input,self.errors,encoding_table)[0]
UnicodeEncodeError: 'charmap' codec can't encode character '\u0107' in position 307: character maps to <undefined>
PS C:\Users\John\Desktop\pdm>
```

To me, it looks like the issue is when it tries to write a non-ascii character to an ascii pipe.

Try:

```
text_body = "\n".join(body)
try:
     ui.echo(text_body, highlight=True)
except UnicodeEncodeError:
    ui.echo(text_body.encode().decode("ascii", errors="ignore"), highlight=True) 
    ui.echo("Problem decoding file as UTF-8.  Some characters may be omit.")
```

Reply:
> Try:

Why, if this fails, no content is produced, then it is no different from just popping the exception info.

Reply:
Ah - for me it _did_ produce content and left a half complete markdown file.  I thought somebody might not notice (e.g. if running in a CI).

I thought that the exception should not stop the other projects being listed - since the charset in the license file is something the pdm user has no control over.